### PR TITLE
Fold wasm-link-metadata into wasm-emscripten-finalize

### PR DESCRIPTION
**Pull request by jgravelle-google**
LLD now emits `__data_end` as a global in the wasm module, meaning we no longer have any information that is only represented in the intermediate wasm object file. Therefore we should merge the two tools and make life a bit simpler.

@sbc100 did most of the C++ work, I did most of the python testing and Emscripten integration.

### CODE REVIEW

File: src/tools/wasm-emscripten-finalize.cpp [hunk 1]
```diff
@@ -76,28 +81,55 @@ int main(int argc, const char *argv[]) {
...
-  // writer.setSymbolMap(symbolMap);
-  writer.setBinary(emitBinary);
-  // if (emitBinary) {
-  //   writer.setSourceMapFilename(sourceMapFilename);
-  //   writer.setSourceMapUrl(sourceMapUrl);
-  // }
-  writer.write(wasm, outfile);
+  if (!emitBinary) {
+    Output output(outfile, Flags::Text, Flags::Release);
+    WasmPrinter::printModule(&wasm, output.getStream());
```
Comment (kripken) [inline src/tools/wasm-emscripten-finalize.cpp]: hmm, why does this manually do the text emitting? `ModuleWriter` can emit both text and binary.

**Comment by aheejin** [comment]
#1395 has conflicts with this, so it looks like this needs merging with master.

File: src/tools/wasm-emscripten-finalize.cpp (hunk 1 above)
Comment (jgravelle-google) [reply src/tools/wasm-emscripten-finalize.cpp]: Because creating an `Output` object to handle the metadata wipes the file if we aren't printing to stdout. It always passes `std::ofstream::trunc`. We could get around this by:
A) adding a trunc/append flag, which is three sets of options to `Output`, which smells like we need a different factoring altogether.
B) modifying `ModuleWriter` to take an `Output` as a parameter to `write`, instead of / in addition to the filename APIs.
C) saying "screw it" and just doing it manually inline.

I went with C, but actually looking at the internals to `ModuleWriter`, it already uses `Output` internally, so it should be pretty trivial to change the interface to take `Output`s as parameters, and refactor the current interface as convenience wrappers around that. That sounds best.

Comment (jgravelle-google) [reply src/tools/wasm-emscripten-finalize.cpp]: Done https://github.com/WebAssembly/binaryen/pull/1408/commits/50d8fa7b4979228aa9473ab7fcfd6920c03b02ad

**Comment by jgravelle-google** [comment]
Ping, any go/no-go?

File: .gitattributes [hunk 2]
```diff
@@ -1 +1,2 @@
-*.a -text
```
Comment (sbc100) [inline .gitattributes]: This is technically unrelated.  Something I did in my original version of this change.

I'm not sure what the old syntax is supposed to do?  Does "-text" work?   Anyway probably want to make this a separate change.

File: check.py [hunk 3]
```diff
@@ -360,7 +360,7 @@ def run_binaryen_js_tests():
     if not s.endswith('.js'): continue
     print s
     f = open('a.js', 'w')
-    binaryen_js = open(os.path.join(options.binaryen_bin, 'binaryen.js')).read()
+    binaryen_js = open(os.path.join(options.binaryen_root, 'bin', 'binaryen.js')).read()
```
Comment (sbc100) [inline check.py]: This is unrelated too, but something i needed to get out-of-tree testing working.

File: scripts/test/generate_lld_tests.py [hunk 4]
```diff
@@ -27,58 +27,54 @@ def files_with_extensions(path, extensions):
       yield file, ext
 
 
-def generate_object_files(clang_bin):
-  print '\n[ building object files from C sources... ]\n'
+def generate_wast_files(clang_bin, lld_bin, emscripten_root):
+  print '\n[ building wasm files from C sources... ]\n'
```
Comment (sbc100) [inline scripts/test/generate_lld_tests.py]: is it wasm or wast?  the print statement and the function name seem to disagree.

File: scripts/test/lld.py [hunk 5]
```diff
@@ -18,36 +18,10 @@
 from support import run_command
 from shared import (
     fail, fail_with_error, files_with_pattern, options,
-    WASM_LINK_METADATA, WASM_EMSCRIPTEN_FINALIZE
+    WASM_EMSCRIPTEN_FINALIZE
```
Comment (sbc100) [inline scripts/test/lld.py]: Does this fit on one line now?

File: scripts/test/lld.py [hunk 6]
```diff
@@ -63,7 +37,8 @@ def test_wasm_emscripten_finalize():
       if ext != '.out' and not os.path.exists(expected_file):
         continue
 
-      cmd = WASM_EMSCRIPTEN_FINALIZE + [wast_path, '-S'] + ext_args
+      cmd = (WASM_EMSCRIPTEN_FINALIZE +
+             [wast_path, '-S', '--global-base=1024'] + ext_args)
```
Comment (sbc100) [inline scripts/test/lld.py]: Can we rely on the default for --global-base rather than passing it here and above?

File: scripts/test/lld.py [hunk 7]
```diff
@@ -75,5 +50,4 @@ def test_wasm_emscripten_finalize():
 
 
 if __name__ == '__main__':
-  test_wasm_link_metadata()
   test_wasm_emscripten_finalize()
```
Comment (sbc100) [inline scripts/test/lld.py]: add `sys.exit(0)` or change to `sys.exit(test_wasm_emscripten_finalize())`?

File: scripts/test/lld.py (hunk 5 above)
Comment (jgravelle-google) [reply scripts/test/lld.py]: 80 chars exactly... and PEP8 specifies <= 79

File: .gitattributes (hunk 2 above)
Comment (jgravelle-google) [reply .gitattributes]: I don't even know what this does, so I'll just revert it

File: check.py (hunk 3 above)
Comment (jgravelle-google) [reply check.py]: Seems harmless to leave in then?

File: scripts/test/lld.py (hunk 6 above)
Comment (jgravelle-google) [reply scripts/test/lld.py]: I'd rather not, we should test that this explicitly sets the value.

... Which means I should explicitly pass something that ISN'T the default to better verify it works.

File: scripts/test/lld.py (hunk 7 above)
Comment (jgravelle-google) [reply scripts/test/lld.py]: None of the other check.py sub-scripts do so I'm inclined to leave it as-is.

File: scripts/test/lld.py (hunk 6 above)
Comment (sbc100) [reply scripts/test/lld.py]: Does emscripten support the non-default path though?  Can the user control this?
If so, then yes I agree should use and test the flag.

Comment (jgravelle-google) [reply scripts/test/lld.py]: Yep and yep https://github.com/kripken/emscripten/pull/6215/files#diff-e64ed58dc838bc8cb95295fef5447e1fR1896

**Comment by kripken** [comment]
This broke CI, flake8 issues on the python changes, e.g. https://travis-ci.org/WebAssembly/binaryen/jobs/341558272

**Comment by jgravelle-google** [comment]
Whoops, sorry https://github.com/WebAssembly/binaryen/pull/1426
